src/render.rs

```diff
@@ -0,0 +1,236 @@
+use matrix_sdk::{
+    events::room::{
+        encrypted::EncryptedEvent,
+        member::{MemberEvent, MembershipChange, MembershipState},
+        message::{
+            AudioMessageEventContent, EmoteMessageEventContent,
+            FileMessageEventContent, ImageMessageEventContent, MessageEvent,
+            MessageEventContent, NoticeMessageEventContent,
+            TextMessageEventContent, VideoMessageEventContent,
+        },
+    },
+    identifiers::UserId,
+};
+use weechat::Weechat;
+
+/// This trait describes events that can be rendered in the weechat UI
+pub(crate) trait RenderableEvent {
+    /// Convert the event into a string that will be displayed in the UI.
+    /// The displayname is taken as a parameter since it cannot be calculated from the event
+    /// context alone.
+    fn render(&self, displayname: &str) -> String;
+
+    /// Get the sender for this event this can be used to get the displayname.
+    fn sender(&self) -> &UserId;
+}
+
+impl RenderableEvent for EncryptedEvent {
+    // TODO: this is not implemented yet
+    fn render(&self, displayname: &str) -> String {
+        let color_user = Weechat::color("green"); // TODO: get per-user color
+        let color_reset = Weechat::color("reset");
+        format!(
+            "{color_user}{}{color_reset}\t{}",
+            displayname,
+            "Unable to decrypt message",
+            color_user = color_user,
+            color_reset = color_reset
+        )
+    }
+
+    fn sender(&self) -> &UserId {
+        &self.sender
+    }
+}
+
+impl RenderableEvent for MemberEvent {
+    fn render(&self, displayname: &str) -> String {
+        use MembershipChange::*;
+        let change_op = self.membership_change();
+        let operation = match change_op {
+            None => "did nothing",
+            Error => "caused an error", // must never happen
+            Joined => "has joined the room",
+            Left => "has left the room",
+            Banned => "was banned",
+            Unbanned => "was unbanned",
+            Kicked => "was kicked from the room",
+            Invited => "was invited to the room",
+            KickedAndBanned => "was kicked and banned",
+            InvitationRejected => "rejected the invitation",
+            InvitationRevoked => "had the invitation revoked",
+            ProfileChanged => "changed the display name or avatar",
+            NotImplemented => "performed an unimplemented operation",
+        };
+
+        let (prefix, color_action) = match change_op {
+            Joined | Invited => ("join", "green"),
+            _ => ("quit", "red"),
```

Comment:
Nick changes have the network prefix on IRC, we should probably do the same.

Reply:
I've handled this, but nick changes are still handled as joins for some reason. I'll take a look at it tomorrow.

Reply:
Fixed.